Avoid DivisionByZeroError

Calvario opened the pull request:
Please give a short description what your pull request is for

Fix #16463 

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [X] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

PipoCanaja commented:
Hi @Calvario 
This fix is indeed necessary. It would be a lot better to also fix the discovery code (where the `discover_sensor` function is called) to ensure it is not called with `0` as `$divisor` parameter.

PipoCanaja commented:
I would start with the sensors of CheckPoint : we need to fix any call with the divisor set to 0. If you run the discovery in "verbose" mode (`./discovery.php -v -d -h <device>`) you should find where the divisor is zero. 

The fix in [includes/discovery/sensors/entity-sensor.inc.php](https://github.com/librenms/librenms/pull/16464/files#diff-130444bfca104534cbbfa4c4fee55fd056c3790528b3f2c0a347a269d15a8b6b) would "mask" any issue and replace them by `$current = 0` which will make troubleshooting difficult if not impossible.

Calvario commented:
I updated the issue, it look's like that it's related with how we handle entPhySensorScale & entPhySensorPrecision.

Indeed this fix would hide the issue but we could add a log or an eventlog to help find out about it.
I already used ./discovery.php in debug and verbose but the device as too many sensors, it make it tedious to check.
I just added (not committed yet) a log inside the "if" to find all the cases.
What do you think ? I think and eventlog is more suitable as we already have the "Error discovering sensors module. Check log file for more details." so it's easier for an user to link them.

As entPhySensorScale "zepto" is not handled, one way to fix it for the current issue would be to add another if like:
```
if ($entry['entPhySensorScale'] == 'zepto') {
    $divisor = '1000000000000000000';
    $multiplier = '1';
}
```

murrant commented:
Improving the multiplier/scale detction would be much better...  Having it default to 1 instead of 0 is ideal :)

Calvario commented:
Like this @murrant ?

murrant commented:
Mostly yes, but I would assign [$divisor, $multiplier] from the match instead of setting it inside the match

murrant commented:
oh, and the multiplier/divisor should be integers, not strings.  (old code set them as strings)

Calvario commented:
Something like that then ?
```
// Try to handle the scale
[$divisor, $multiplier] = match ($entry['entPhySensorScale']) {
    'zepto' => [1000000000000000000, 1],
    'nano' => [1000000000, 1],
    'micro' => [1000000, 1],
    'milli' => [1000, 1],
    'units' => [1, 1],
    'kilo' => [1, 1000],
    'mega' => [1, 1000000],
    'giga' => [1, 1000000000],
    'yocto' => [1, 1],
    default => [1, 1],
};
```

murrant commented on includes/discovery/sensors/entity-sensor.inc.php:
```diff
@@ -112,41 +112,23 @@
+            };
+
+            if (is_numeric($entry['entPhySensorPrecision']) && $entry['entPhySensorPrecision'] > 0) {
+                $divisor .= str_pad('', $entry['entPhySensorPrecision'], '0');
```
This code is actually duplicated in cisco-entity-sensors too :D

murrant replied:
Too many questions :)  Mostly, this code base has a lot of legacy.
